modules/seqexec/server/src/main/scala/seqexec/server/ObserveActions.scala

```diff
@@ -30,84 +30,86 @@ trait ObserveActions {
-    )
-
-  private def sendDataEnd[F[_]: Monad](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId, dataId: String): SeqActionF[F, Unit] =
-    systems.odb.datasetComplete(obsId, dataId, imageFileId).ifM(
-      SeqActionF.void,
-      SeqActionF.raiseException(SeqexecFailure.Unexpected("Unable to send DataEnd message to ODB.")))
-
-  def observe[F[_]: Sync](systems: Systems[F], config: Config, obsId: Observation.Id, inst: InstrumentSystem[F],
-                      otherSys: List[System[F]], headers: Reader[HeaderExtraData, List[Header[F]]])
-                     (ctx: HeaderExtraData)
-                     (implicit ev: Concurrent[F]): Stream[F, Result[F]]
+  private def dhsFileId[F[_]: ApplicativeError[?[_], Throwable]](systems: Systems[F], inst: InstrumentSystem[F]): F[ImageFileId] =
+    systems.dhs.createImage(DhsClient.ImageParameters(DhsClient.Permanent, List(inst.contributorName, "dhs-http")))
+
+  private def info[F[_]: Sync](msg: => String): F[Unit] = Sync[F].delay(Log.info(msg))
+
+  private def abortTail[F[_]: MonadError[?[_], Throwable]](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId): F[Result[F]] =
+    systems.odb.obsAbort(obsId, imageFileId)
+      .ensure(SeqexecFailure.Unexpected("Unable to send ObservationAborted message to ODB."))(identity) *>
+    MonadError[F, Throwable].raiseError(SeqexecFailure.Execution(s"Observation ${obsId.format} aborted by user."))
+
+  private def sendDataStart[F[_]: MonadError[?[_], Throwable]](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId, dataId: String): F[Unit] =
+    systems.odb.datasetStart(obsId, dataId, imageFileId)
+      .ensure(SeqexecFailure.Unexpected("Unable to send DataStart message to ODB."))(identity)
+      .void
+
+  private def sendDataEnd[F[_]: MonadError[?[_], Throwable]](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId, dataId: String): F[Unit] =
+    systems.odb.datasetComplete(obsId, dataId, imageFileId)
+      .ensure(SeqexecFailure.Unexpected("Unable to send DataEnd message to ODB."))(identity)
+      .void
+
+  def observe[F[_]: Sync: Concurrent](
+    systems: Systems[F],
+    config: Config,
+    obsId: Observation.Id,
+    inst: InstrumentSystem[F],
+    otherSys: List[System[F]],
+    headers: Reader[HeaderExtraData, List[Header[F]]])(
+      ctx: HeaderExtraData): Stream[F, Result[F]]
   = {
-    def dataId: SeqActionF[F, String] = SeqActionF.either(
-      config.extractAs[String](OBSERVE_KEY / DATA_LABEL_PROP).leftMap(e =>
-      SeqexecFailure.Unexpected(ConfigUtilOps.explain(e))))
+    def dataId: F[String] =
+      EitherT.fromEither[F](
+        config.extractAs[String](OBSERVE_KEY / DATA_LABEL_PROP)
+          .leftMap(e => SeqexecFailure.Unexpected(ConfigUtilOps.explain(e)))
+      ).widenRethrowT
 
-    def notifyObserveStart: SeqActionF[F, Unit] =
+    def notifyObserveStart: F[Unit] =
       otherSys.map(_.notifyObserveStart).sequence.void
```

Comment:
`otherSys.traverse_(_.notifyObserveStart)` here and elsewhere. I know it's not what you're working on but can we simplify it anyway?

Reply:
👍 A question, why `traverse_` rather than `traverse`?